tools/chplcheck/src/config.py

```diff
@@ -18,8 +18,54 @@
 #
 
 from dataclasses import dataclass
-from typing import List, Union, Dict, Any
+from typing import List, Union, Dict, Any, Optional
 import argparse
+import re
+
+
+@dataclass(frozen=True)
+class RuleSetting:
+    """
+    Settings for an rules, these can optionally be namespaced to a specific rule
+    """
+
+    setting_name: str
+    rule_name: Optional[str] = None
+
+    def __str__(self):
+        return (
+            f"{self.rule_name}.{self.setting_name}"
+            if self.rule_name
+            else self.setting_name
+        )
+
+
+class RuleSettingAction(argparse.Action):
+
+    def __init__(self, *args, **kwargs):
+        if "default" not in kwargs:
+            kwargs["default"] = dict()
+        elif not isinstance(kwargs["default"], dict):
+            raise ValueError("default must be a dict")
+        super().__init__(*args, **kwargs)
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        # values must match the following regex, if they don't raise an error
+        regex = r"^(\w+\.)?(\w+)=(.+)$"
+        if not isinstance(values, str):
+            raise argparse.ArgumentTypeError(f"Invalid rule setting: {values}")
+        m = re.match(regex, values)
+        if not m:
+            raise argparse.ArgumentTypeError(f"Invalid rule setting: {values}")
+
+        rule_name = m.group(1)[:-1] if m.group(1) else None
+        setting_name = m.group(2)
+        setting = RuleSetting(setting_name, rule_name)
+        setting_value = m.group(3)
+
+        setting_dict = getattr(namespace, self.dest)
+        setting_dict[setting] = setting_value
+        setattr(namespace, self.dest, setting_dict)
```

Comment:
do you need the setattr? Seems like the object of the attribute is still the same one (the dict), so modifying the dict should be sufficient. I'm not familiar with namespaces, though.

Reply:
its should be equivalent to `namespace[self.dest] = setting_dict`, which I think also works